src/Client.cpp

```diff
@@ -966,6 +997,149 @@ bool CClient::OnActionMessage(CActionMessage& Message) {
     return true;
 }
 
+void CClient::OnAuthenticateMessage(CAuthenticateMessage& Message) {
+    const auto uiMaxSASLMsgLength = 400u;
+    auto bAuthenticationSuccess = false;
+    auto sMessage = Message.GetText();
+    const auto iBufferSize = sMessage.length();
+
+    auto SASLReset = [this]() {
+        m_sSASLMechanism = "";
+        m_sSASLBuffer = "";
+    };
+
+    auto SASLChallenge = [this](CString sChallenge) {
+        sChallenge.Base64Encode();
+        auto sChallengeSize = sChallenge.length();
+
+        if (sChallengeSize > uiMaxSASLMsgLength) {
+            for (int i = 0; i < sChallengeSize; i += uiMaxSASLMsgLength) {
+                CString sMsgPart = sChallenge.substr(i, uiMaxSASLMsgLength);
+                PutClient("AUTHENTICATE " + sMsgPart);
+            }
+        } else if (sChallengeSize > 0) {
+            PutClient("AUTHENTICATE " + sChallenge);
+        }
+        if (sChallengeSize % uiMaxSASLMsgLength == 0) {
+            PutClient("AUTHENTICATE +");
+        }
+    };
+
+    if (!m_bSASL) return;
+	
+	if (!m_sSASLUser.empty() || IsAttached()) {
+        PutClient(":irc.znc.in 907 " + GetNick() +
+                  " :You have already authenticated using SASL");
+        return;
+    }
+    
+	if (!m_bSASLAuthenticating || sMessage.Equals("*")) {
+        PutClient(":irc.znc.in 906 " + GetNick() +
+                  " :SASL authentication aborted");
+        if (!IsAttached()) {
+			m_bSASLAuthenticating = false;
+            SASLReset();
+		}
+        return;
+    }
+
+	if (iBufferSize > uiMaxSASLMsgLength) {
+        PutClient(":irc.znc.in 905 " + GetNick() + " :SASL message too long");
+        SASLReset();
+        return;
+    }
+
+	if (m_sSASLMechanism.empty()) {
+        SCString ssMechanisms;
+        auto sMechanisms = EnumerateSASLMechanisms(ssMechanisms);
+
+        if (ssMechanisms.find(sMessage) == ssMechanisms.end()) {
+            PutClient(":irc.znc.in 908 " + GetNick() + " " + sMechanisms +
+                      " :are available SASL mechanisms");
+            PutClient(":irc.znc.in 904 " + GetNick() +
+                      " :SASL authentication failed");
+            SASLReset();
+
+            return;
+        }
+
+        m_sSASLMechanism = sMessage;
+
+        auto bResult = false;
+        CString sChallenge;
+        GLOBALMODULECALL(OnSASLServerChallenge(m_sSASLMechanism, sChallenge),
+                         &bResult);
+        if (bResult) {
+            SASLChallenge(sChallenge);
+        } else {
+            PutClient("AUTHENTICATE +");
+        }
+		return;
+	}
+
+    if (m_sSASLBuffer.length() + sMessage.length() > 10 * 1024) {
+        PutClient(":irc.znc.in 904 " + GetNick() + " :SASL response too long");
+        SASLReset();
+        return;
+    }
```

Comment:
Could you reformat the code? It's hard to read and I don't understand what's going on